### PR TITLE
SALTO-3829:  Support references to transitions

### PR DESCRIPTION
**Pull request by IdoZakk**
Support references in scriptRunner for transitions

---

The references are done to the key in the map
Before deploy we guess the future transition ids based on the logic confirmed by Atlassian.
After the deployment we check to see if the guess was successful, if not we deploy again with the ids we got from the deployment under the assumption that the ids will be the same the second time. If that is not the case we fail the deployment.

To achieve the changes in the deployment I did some refactoring of the code. 
* The external call for transitions ids was taken out of trigger deployment
* The ids are now added to the transitions after deployment, and deleted in the transitionIds filter (to allow the workflow references filter to make the proper references)
* Some logic was taken out to functions so it can be reused

---
_Release Notes_: 
Jira Adapter: 
* ScriptRunner: Transitions ids fields are now turned into references and are updated during deployment

---
_User Notifications_: 
Jira Adapter: 
* ScriptRunner: Transitions ids fields are now turned into references

### CODE REVIEW

File: packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts [hunk 1]
```diff
@@ -13,12 +13,13 @@
 * See the License for the specific language governing permissions and
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts:13]: Are the changes here just a refactor of the changes of the previous PR or are they related to the transition reference

**Review by alonstern — COMMENTED** [review]
🔥

File: packages/jira-adapter/src/filters/script_runner/workflow/workflow_references.ts [hunk 2]
```diff
@@ -15,30 +15,84 @@
...
       }
-      await awu(changes)
+      const workflows = changes
         .filter(isAdditionOrModificationChange)
         .map(getChangeData)
         .filter(isInstanceElement)
-        .filter(instance => instance.elemID.typeName === WORKFLOW_TYPE_NAME)
+        .filter(isWorkflowInstance)
+
+      workflows
```
[the remark this answers is not in the document]
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/script_runner/workflow/workflow_references.ts:83]: The problem is we have to encode the entire ScriptRunner Object, so when we get to `resolveValues` it is too late

File: packages/jira-adapter/src/filters/workflow/transition_structure.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
+/*
```
[the remark this answers is not in the document]
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/workflow/transition_structure.ts:1]: Just moving the code

File: packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts [hunk 4]
```diff
@@ -38,6 +41,53 @@ export const INITIAL_VALIDATOR = {
   },
 }
 
+const isValidTransitionResponse = (response: unknown): response is { values: [WorkflowResponse] } => {
```
[the remark this answers is not in the document]
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts:44]: It was a copy paste from Tirggers, do you think I should change?

File: packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts [hunk 5]
```diff
@@ -82,14 +132,104 @@ const workflowTransitionsToList = (workflowInstance: InstanceElement): void => {
...
 export const deployWorkflow = async (
   change: AdditionChange<InstanceElement> | RemovalChange<InstanceElement>,
   client: JiraClient,
   config: JiraConfig,
 ): Promise<void> => {
   const resolvedChange = await resolveChangeElement(change, getLookUpName)
 
-  if (!isPostFetchWorkflowChange(resolvedChange)) {
+  if (!isPostFetchWorkflowChange(resolvedChange)
+    || !isPostFetchWorkflowChange(change)) {
```
[the remark this answers is not in the document]
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts:232]: Without it I get type errors later on as change is not recognized as a workflow change

File: packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts (hunk 1 above)
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts:13]: refactor

**Review by alonstern — APPROVED** [review]
https://github.com/salto-io/salto/pull/4705 Was merged 👍 Could you just check the deploy and fetch all work with circle before merging?

**Pull request merged**